packages/cubejs-redshift-driver/src/RedshiftDriver.ts

```diff
@@ -5,14 +5,24 @@ import crypto from 'crypto';
 import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
 import { S3, GetObjectCommand } from '@aws-sdk/client-s3';
 
-interface RedshiftDriverExportAWS {
+interface RedshiftDriverExportRequiredAWS {
   bucketType: 's3',
   bucketName: string,
-  keyId: string,
-  secretKey: string,
   region: string,
 }
 
+interface RedshiftDriverExportArnAWS extends RedshiftDriverExportRequiredAWS{
+  unloadArn?: string,
+}
+
+interface RedshiftDriverExportKeySecretAWS extends RedshiftDriverExportRequiredAWS{
+  keyId?: string,
+  secretKey?: string,
+}
+
+interface RedshiftDriverExportAWS extends RedshiftDriverExportArnAWS, RedshiftDriverExportKeySecretAWS {
```

Comment:
Can we have *both* key/secret and ARN? Which one takes precedence?

If they are mutually exclusive, then can you please use discriminated unions (and no optional fields) along the lines of:

```
interface RedshiftDriverExportArnAWS extends RedshiftDriverExportRequiredAWS {
  type: 'arn',
  unloadArn: string,
}

interface RedshiftDriverExportKeySecretAWS extends RedshiftDriverExportRequiredAWS {
  keyId: string,
  secretKey: string,
}

type RedshiftDriverExportAWS = RedshiftDriverExportArnAWS | RedshiftDriverExportKeySecretAWS
```

Reply:
I spent quite a bit figuring out how to do this... and failed!  Thanks for the pointer, would definitely make it easier to read.  Should be a similar option for when both are present as well.

Reply:
As for your question, it turns out to be a two parter.

- For the `unload` operation in redshift, this proposed solution _prefers_ the `unloadArn` if both are present.
- It is my assumption, that upstream in the codebase, there is a configuration block for `S3Client` that loads the key/secret into its configuration.  If they are `undefined`, the underlying `S3Client` SDK will use the `iam_role` that is assigned to whatever is running it.  If they are **both** defined it will use the provided key/secret pair.  I am uncertain as to what will happen if only one of the API Key parameters is present.